// @keycloak/keycloak-admin-client/lib/resources/components.js
import Resource from './resource.js';
export class Components extends Resource {
    constructor(client) {
        super(client, {
            path: '/admin/realms/{realm}/components',
            getUrlParams: () => ({
                realm: client.realmName,
            }),
            getBaseUrl: () => client.baseUrl,
        });
        this.find = this.makeRequest({
            method: 'GET',
        });
        this.create = this.makeRequest({
            method: 'POST',
            returnResourceIdInLocationHeader: { field: 'id' },
        });
        this.findOne = this.makeRequest({
            method: 'GET',
            path: '/{id}',
            urlParamKeys: ['id'],
            catchNotFound: true,
        });
        this.update = this.makeUpdateRequest({
            method: 'PUT',
            path: '/{id}',
            urlParamKeys: ['id'],
        });
        this.del = this.makeRequest({
            method: 'DELETE',
            path: '/{id}',
            urlParamKeys: ['id'],
        });
        this.listSubComponents = this.makeRequest({
            method: 'GET',
            path: '/{id}/sub-component-types',
            urlParamKeys: ['id'],
            queryParamKeys: ['type'],
        });
    }
}
