// @keycloak/keycloak-admin-client/lib/resources/userStorageProvider.js
import Resource from './resource.js';
export class UserStorageProvider extends Resource {
    constructor(client) {
        super(client, {
            path: '/admin/realms/{realm}/user-storage',
            getUrlParams: () => ({
                realm: client.realmName,
            }),
            getBaseUrl: () => client.baseUrl,
        });
        this.name = this.makeRequest({
            method: 'GET',
            path: '/{id}/name',
            urlParamKeys: ['id'],
        });
        this.removeImportedUsers = this.makeRequest({
            method: 'POST',
            path: '/{id}/remove-imported-users',
            urlParamKeys: ['id'],
        });
        this.sync = this.makeRequest({
            method: 'POST',
            path: '/{id}/sync',
            urlParamKeys: ['id'],
            queryParamKeys: ['action'],
        });
        this.unlinkUsers = this.makeRequest({
            method: 'POST',
            path: '/{id}/unlink-users',
            urlParamKeys: ['id'],
        });
        this.mappersSync = this.makeRequest({
            method: 'POST',
            path: '/{parentId}/mappers/{id}/sync',
            urlParamKeys: ['id', 'parentId'],
            queryParamKeys: ['direction'],
        });
    }
}
